ql/src/java/org/apache/hadoop/hive/ql/exec/tez/SplitGrouper.java

```diff
@@ -71,7 +71,7 @@ public class SplitGrouper {
 
   // TODO This needs to be looked at. Map of Map to Map... Made concurrent for now since split generation
   // can happen in parallel.
-  private static final Map<Map<Path, PartitionDesc>, Map<Path, PartitionDesc>> cache =
+  private final Map<Map<Path, PartitionDesc>, Map<Path, PartitionDesc>> cache =
```

Comment:
Are you using Hive4-beta?
If it is possible, can we add a test?

Reply:
I've looked through the code, and the way I've done it doesn't change any of the logical behaviour of the code. I just defined the cache object as a non-static property to avoid the scenario of multiple threads calling it at the same time.

Reply:
况且,如果遵循原本的定义方式,当多个线程同时访问cache,那么有很大概率会出现死锁.

Reply:
Since the nature of the problem is that HashTable causes deadlocks in concurrent scenarios, I'm not sure what kind of test case would be appropriate to add ....

Reply:
@zhangbutao  If there's anything else I can provide, please let me know.

Reply:
@BsoBird, with that change the whole caching construct is useless when we have multiple SplitGrouper instances. 
The issue is with concurrent Hashtable.equals(). What if we change the type of PartitionDesc.properties to HashMap?

Reply:
@deniskuzZ 
hi. I think,modified the Properties to HashMap is too expensive, because HashMap does not have the Properties class of the various related functions. For example: loadFromXML ().
But I agree with you that there are some advantages to be lost by making the cache non-static. So I have a second option:
org.apache.hadoop.hive.common.CopyOnFirstWriteProperties::equals
```
  @Override
  public synchronized boolean equals(Object o) {
    if(o instanceof Hashtable){
      o = new Hashtable<>((Hashtable<Object, Object>) o);
    }
    if (interned != null) return interned.equals(o);
    else return super.equals(o);
  }
```

Reply:
@deniskuzZ 
The third option is exactly the one you described:
PartitionDesc.equals::
```
    Properties properties1 = getProperties();
    Properties properties2 = other.getProperties();
    cond = (properties1 == null && properties2 == null) ||
        (properties1 != null && Maps.difference(properties1,properties2).areEqual());
    if (!cond) {
      return false;
    }
```

Reply:
> if removing static, multiple SplitGrouper instances will have each own cache; I am afraid of this will consume much memory of LLAP coordinator (Not sure i understand correctlly).
> 
> From the jstack, it seems that thread is blocked at `org.apache.hadoop.hive.common.CopyOnFirstWriteProperties::equals`, which was introduced to resolve high memory usage about `Properties `, [HIVE-16079](https://issues.apache.org/jira/browse/HIVE-16079). Maybe we can find more clues on `CopyOnFirstWriteProperties`? Just my subjective thought, for your information. Thanks.

@zhangbutao  Since CopyOnFirstWriteProperties is a proxy class for Properties and implements the HashTable interface, CopyOnFirstWriteProperties also triggers deadlocks in concurrent scenarios. You can see the code.

Reply:
@deniskuzZ Hi, what do you think is the best way for us to proceed?

Reply:
not sure why we need to check properties in `equals`, but the same is done in `TableDesc`. In my opinion, non-synchronized equals should be enough.

Reply:
@deniskuzZ Okay. let me resubmit a PR.

Reply:
@deniskuzZ We can use HashMap instead of ConcurrentHashMap, right?

Reply:
![image](https://github.com/apache/hive/assets/30249475/1dbb598f-f355-490a-919b-9a28490b7650)

Reply:
yes,
@BsoBird , could you please add q test that triggers use of cache from the SplitGrouper.

Reply:
@deniskuzZ   I'm a newbie in this field, for a long time I can't even debug HIVE code locally(because I always use windows OS computer). I would be grateful if you could guide me. Tks.

Reply:
basically you can add the sql into q file to replicate the use-case and define the expectations in the q.out file
more details on how to execute those tests:
https://cwiki.apache.org/confluence/display/hive/hivedeveloperfaq#HiveDeveloperFAQ-HowdoIruntheclientpositive/clientnegativeunittests?

Reply:
add the SQL you've used to repro the concurrency issue into that q.file
note, in case of iceberg, place the q file under `itests/qtest-iceberg`
{code}
mvn test -Dtest=TestIcebergLlapLocalCliDriver -Dqfile=mytest.q -Drat.skip=true -Dtest.output.overwrite=true
{code}

Reply:
@deniskuzZ Hi. @deniskuzZ . My current problem is that I can't reproduce the problem consistently. I haven't worked out a way to reproduce the problem consistently yet. So I have no way to add a q.file test case at the moment.

Reply:
@BsoBird, could you please share the repro steps that led to the issue? it doesn't have to be consistent. I just want to create at least one test covering that part of the code.  At this moment we have 0 coverage.

Reply:
@deniskuzZ  Just Read IcebergTable. and using hive llap.  Before I used this patch, this problem occurred occasionally. After I used this patch, the problem disappeared.

Reply:
@deniskuzZ If it is difficult to override this change with SQL, can we re-run the previous SQL to test the LLAP+ICEBERG reads and writes? Because our change must affect all the SQL that reads and writes to the ICEBERG table, we just need to verify that the output is the same as before.